File: server/use-cases/getActions.ts
import prisma from "../utils/prisma"

import {
    GetActionsParamsType,
} from '../../schemas/preventive'

export async function getActions(input: GetActionsParamsType) {
    try {
        const { machineId, natureId, searchText, weekCode, showIgnore } = input
        const nextExecution = weekCode
        const actions = await prisma.preventiveAction.findMany({
            skip: input.cursor === 1 ? 0 : 1,
            take: input.limit,
            cursor: { id: input?.cursor },
            orderBy: { id: 'asc' },
            where: {
                OR: {
                    description: { contains: searchText },
                },
                ...machineId >= 0 ? { machineId } : {},
                ...natureId >= 0 ? { natureId } : {},
                ...nextExecution != '' ? { nextExecution } : {},
                ...showIgnore ? {} : { ignore: false }
            },
            include: {
                nature: true, machine: true, _count: { select: { actionsTaken: true } }
            }
        })
        return actions
    } catch (error) {
        throw error
    }
}

File: src-py/use_cases/get_actions.py
from schemas.preventive import GetActionsParams
from infra.prisma import prisma


async def getActions(params: dict) -> list[dict]:
    validated = GetActionsParams.model_validate(params)

    next_execution = validated.weekCode

    where: dict = {
        "OR": [{"description": {"contains": validated.searchText}}],
    }
    if validated.machineId >= 0:
        where["machineId"] = validated.machineId
    if validated.natureId >= 0:
        where["natureId"] = validated.natureId
    if next_execution != "":
        where["nextExecution"] = next_execution
    if not validated.showIgnore:
        where["ignore"] = False

    query: dict = {
        "skip": 0 if validated.cursor == 1 else 1,
        "take": validated.limit,
        "order": {"id": "asc"},
        "where": where,
        "include": {
            "nature": True,
            "machine": True,
            "actionsTaken": True,
        },
    }
    if validated.cursor is not None:
        query["cursor"] = {"id": validated.cursor}

    actions = await prisma.preventiveaction.find_many(**query)

    result: list[dict] = []
    for action in actions:
        data = action.model_dump(mode="json")
        count = len(data.pop("actionsTaken", []) or [])
        data["_count"] = {"actionsTaken": count}
        result.append(data)

    return result
